File: app/src/main/java/com/josegonzalez/musicsearchsample/view/itemList/CustomAdapter.java
package com.josegonzalez.musicsearchsample.view.itemList;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.josegonzalez.musicsearchsample.R;
import com.josegonzalez.musicsearchsample.model.TrackDto;

import java.util.List;

public class CustomAdapter extends RecyclerView.Adapter<CustomAdapter.TrackViewHolder> {

    private List<TrackDto> dataSet;
    private TrackAdapterCallback trackAdapterCallback;

    private Context context;

    public interface TrackAdapterCallback{
        void onListItemClicked(int position);
    }

    public static class TrackViewHolder extends RecyclerView.ViewHolder {
        TextView textViewName;
        TextView textViewVersion;
        ImageView imageViewIcon;

        public TrackViewHolder(View itemView) {
            super(itemView);
            this.textViewName = (TextView) itemView.findViewById(R.id.textViewName);
            this.textViewVersion = (TextView) itemView.findViewById(R.id.textViewVersion);
            this.imageViewIcon = (ImageView) itemView.findViewById(R.id.imageView);
        }
    }

    public CustomAdapter(List<TrackDto> data, TrackAdapterCallback trackAdapterCallback) {
        this.dataSet = data;
        this.trackAdapterCallback = trackAdapterCallback;
    }

    @Override
    public TrackViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext()).inflate(R.layout.cards_layout, parent, false);
        context = parent.getContext();
        TrackViewHolder myViewHolder = new TrackViewHolder(view);
        myViewHolder.itemView.setOnClickListener(v ->
                trackAdapterCallback.onListItemClicked(myViewHolder.getAdapterPosition()));
        return myViewHolder;
    }

    @Override
    public void onBindViewHolder(final TrackViewHolder holder, final int listPosition) {
        TextView textViewName = holder.textViewName;
        TextView textViewVersion = holder.textViewVersion;
        ImageView imageView = holder.imageViewIcon;

        textViewName.setText(dataSet.get(listPosition).getTrackName());
        textViewVersion.setText(dataSet.get(listPosition).getArtistName());

        if (dataSet.get(listPosition).getArtworkUrl100() != null && !dataSet.get(listPosition).getArtworkUrl100().isEmpty()) {
            Glide.with(context)
                    .load(dataSet.get(listPosition).getArtworkUrl100())
                    .into(imageView);
        }
    }

    @Override
    public int getItemCount() {
        return dataSet.size();
    }

}

File: app/src/main/java/com/josegonzalez/musicsearchsample/features/list/CustomAdapter.kt
package com.josegonzalez.musicsearchsample.features.list

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bumptech.glide.Glide
import com.josegonzalez.musicsearchsample.R
import com.josegonzalez.musicsearchsample.model.TrackDto
import com.josegonzalez.musicsearchsample.features.list.CustomAdapter.TrackViewHolder

class CustomAdapter(private val dataSet: List<TrackDto>, private val trackAdapterCallback: TrackAdapterCallback) : RecyclerView.Adapter<TrackViewHolder>() {
    private var context: Context? = null

    interface TrackAdapterCallback {
        fun onListItemClicked(position: Int)
    }

    class TrackViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        var textViewName: TextView = itemView.findViewById<View>(R.id.textViewName) as TextView
        var textViewVersion: TextView = itemView.findViewById<View>(R.id.textViewVersion) as TextView
        var imageViewIcon: ImageView = itemView.findViewById<View>(R.id.imageView) as ImageView

    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): TrackViewHolder {
        val view = LayoutInflater.from(parent.context).inflate(R.layout.cards_layout, parent, false)
        context = parent.context
        val myViewHolder = TrackViewHolder(view)
        myViewHolder.itemView.setOnClickListener {
            trackAdapterCallback.onListItemClicked(myViewHolder.adapterPosition)
        }
        return myViewHolder
    }

    override fun onBindViewHolder(holder: TrackViewHolder, listPosition: Int) {
        val textViewName = holder.textViewName
        val textViewVersion = holder.textViewVersion
        val imageView = holder.imageViewIcon
        textViewName.text = dataSet[listPosition].trackName
        textViewVersion.text = dataSet[listPosition].artistName
        if (dataSet[listPosition].artworkUrl100 != null && !dataSet[listPosition].artworkUrl100?.isEmpty()!!) {
            context?.let {
                Glide.with(it)
                        .load(dataSet[listPosition].artworkUrl100)
                        .into(imageView)
            }
        }
    }

    override fun getItemCount(): Int {
        return dataSet.size
    }

}